File: lib/data/export.dart
import 'dart:convert';
import 'dart:io';

import 'package:path_provider/path_provider.dart';
import 'package:pdf/pdf.dart';
import 'package:pdf/widgets.dart' as pw;
import 'package:share_plus/share_plus.dart';

import '../format.dart';
import '../models.dart';

String _stamp() => todayIso();

String _csvCell(Object? value) {
  final s = value?.toString() ?? '';
  if (RegExp(r'["\n\r,]').hasMatch(s)) return '"${s.replaceAll('"', '""')}"';
  return s;
}

String _typeLabel(String type) => typeMeta[type]?.label ?? type;

String slug(String? name) {
  final base = (name ?? 'export')
      .toLowerCase()
      .replaceAll(RegExp(r'[^a-z0-9]+'), '-')
      .replaceAll(RegExp(r'^-|-$'), '');
  final trimmed = base.length > 40 ? base.substring(0, 40) : base;
  return trimmed.isEmpty ? 'export' : trimmed;
}

class _LogWithVehicle {
  final LogEntry log;
  final Vehicle vehicle;
  _LogWithVehicle(this.log, this.vehicle);
}

List<_LogWithVehicle> _allLogs(List<Vehicle> vehicles) {
  final all = vehicles.expand((v) => v.logs.map((l) => _LogWithVehicle(l, v))).toList();
  all.sort((a, b) => DateTime.parse(b.log.date).compareTo(DateTime.parse(a.log.date)));
  return all;
}

String buildJson(List<Vehicle> vehicles) {
  return const JsonEncoder.withIndent('  ').convert({
    'app': 'GreaseTrail',
    'version': 1,
    'exportedAt': DateTime.now().toIso8601String(),
    'vehicles': vehicles.map((v) => v.toJson()).toList(),
  });
}

String buildCsv(List<Vehicle> vehicles) {
  final header = ['Vehicle', 'Type', 'Model', 'Date', 'Odometer_km', 'Category', 'Service', 'Values', 'Note'];
  final rows = _allLogs(vehicles).map((e) {
    final l = e.log;
    final v = e.vehicle;
    return [
      v.name,
      _typeLabel(v.type),
      v.model,
      l.date,
      l.odometer,
      l.categoryName,
      l.service ?? '',
      formatLogLine(v, l),
      l.note,
    ].map(_csvCell).join(',');
  });
  return [header.join(','), ...rows].join('\n');
}

pw.Document buildPdfDocument(List<Vehicle> vehicles) {
  final doc = pw.Document();
  final single = vehicles.length == 1;
  final title = single ? vehicles.first.name : 'GreaseTrail';
  final sub = single
      ? [
          formatDate(todayIso()),
          vehicles.first.model,
          _typeLabel(vehicles.first.type),
          '${formatOdo(vehicles.first.odometer)} km',
        ].where((s) => s.isNotEmpty).join('  ·  ')
      : '${formatDate(todayIso())}  ·  ${vehicles.length} vehicles';

  doc.addPage(
    pw.MultiPage(
      build: (context) => [
        pw.Text(title.toUpperCase(), style: pw.TextStyle(fontSize: 20, fontWeight: pw.FontWeight.bold, letterSpacing: 1.5)),
        pw.SizedBox(height: 4),
        pw.Text(sub, style: const pw.TextStyle(fontSize: 10, color: PdfColor.fromInt(0xFF8A8A84))),
        pw.SizedBox(height: 20),
        if (vehicles.isEmpty) pw.Text('No vehicles.', style: const pw.TextStyle(color: PdfColor.fromInt(0xFF8A8A84))),
        for (final v in vehicles) ...[
          if (!single) ...[
            pw.SizedBox(height: 14),
            pw.Text(v.name.toUpperCase(), style: pw.TextStyle(fontSize: 14, fontWeight: pw.FontWeight.bold)),
            pw.Text(
              [v.model, _typeLabel(v.type), '${formatOdo(v.odometer)} km'].where((s) => s.isNotEmpty).join('  ·  '),
              style: const pw.TextStyle(fontSize: 10, color: PdfColor.fromInt(0xFF8A8A84)),
            ),
            pw.SizedBox(height: 8),
          ],
          if (v.logs.isEmpty)
            pw.Text('No log entries.', style: const pw.TextStyle(color: PdfColor.fromInt(0xFF8A8A84)))
          else
            pw.Table(
              border: const pw.TableBorder(horizontalInside: pw.BorderSide(color: PdfColor.fromInt(0xFFE4E2DC))),
              columnWidths: const {
                0: pw.FlexColumnWidth(2),
                1: pw.FlexColumnWidth(1.4),
                2: pw.FlexColumnWidth(2.4),
                3: pw.FlexColumnWidth(4),
              },
              children: [
                pw.TableRow(children: [
                  for (final h in ['Date', 'km', 'Category', 'Details'])
                    pw.Padding(
                      padding: const pw.EdgeInsets.only(bottom: 6),
                      child: pw.Text(h.toUpperCase(),
                          style: pw.TextStyle(
                              fontSize: 8, fontWeight: pw.FontWeight.bold, color: const PdfColor.fromInt(0xFF8A8A84))),
                    ),
                ]),
                for (final l in [...v.logs]..sort((a, b) => DateTime.parse(b.date).compareTo(DateTime.parse(a.date))))
                  pw.TableRow(children: [
                    pw.Padding(padding: const pw.EdgeInsets.symmetric(vertical: 4), child: pw.Text(formatDate(l.date), style: const pw.TextStyle(fontSize: 9))),
                    pw.Padding(padding: const pw.EdgeInsets.symmetric(vertical: 4), child: pw.Text(formatOdo(l.odometer), style: const pw.TextStyle(fontSize: 9))),
                    pw.Padding(padding: const pw.EdgeInsets.symmetric(vertical: 4), child: pw.Text(l.categoryName, style: const pw.TextStyle(fontSize: 9))),
                    pw.Padding(
                      padding: const pw.EdgeInsets.symmetric(vertical: 4),
                      child: pw.Column(
                        crossAxisAlignment: pw.CrossAxisAlignment.start,
                        children: [
                          if (l.service != null) pw.Text(l.service!, style: pw.TextStyle(fontSize: 9, fontWeight: pw.FontWeight.bold)),
                          if (formatLogLine(v, l).isNotEmpty) pw.Text(formatLogLine(v, l), style: const pw.TextStyle(fontSize: 9)),
                          if (l.note.isNotEmpty)
                            pw.Text(l.note, style: const pw.TextStyle(fontSize: 9, color: PdfColor.fromInt(0xFF8A8A84))),
                          if (l.service == null && formatLogLine(v, l).isEmpty && l.note.isEmpty)
                            pw.Text('—', style: const pw.TextStyle(fontSize: 9)),
                        ],
                      ),
                    ),
                  ]),
              ],
            ),
        ],
      ],
    ),
  );
  return doc;
}

Future<File> _writeCache(String name, List<int> bytes) async {
  final dir = await getTemporaryDirectory();
  final file = File('${dir.path}/$name');
  await file.writeAsBytes(bytes);
  return file;
}

Future<void> _shareFile(File file, {required String mimeType, String? subject}) async {
  await Share.shareXFiles([XFile(file.path, mimeType: mimeType)], subject: subject);
}

Future<void> exportJson(List<Vehicle> vehicles) async {
  final file = await _writeCache('greasetrail-${_stamp()}.json', utf8.encode(buildJson(vehicles)));
  await _shareFile(file, mimeType: 'application/json', subject: 'Export JSON');
}

Future<void> exportCsv(List<Vehicle> vehicles) async {
  final file = await _writeCache('greasetrail-${_stamp()}.csv', utf8.encode(buildCsv(vehicles)));
  await _shareFile(file, mimeType: 'text/csv', subject: 'Export CSV');
}

Future<void> exportPdf(List<Vehicle> vehicles) async {
  final label = vehicles.length == 1 ? slug(vehicles.first.name) : 'all';
  final bytes = await buildPdfDocument(vehicles).save();
  final file = await _writeCache('greasetrail-$label-${_stamp()}.pdf', bytes);
  await _shareFile(file, mimeType: 'application/pdf', subject: 'Export PDF');
}

Future<void> shareJsonFile(String filename, String contents, String subject) async {
  final file = await _writeCache(filename, utf8.encode(contents));
  await _shareFile(file, mimeType: 'application/json', subject: subject);
}

File: src/data/export.js
import { Alert } from "react-native";
import * as FileSystem from "expo-file-system";
import * as Sharing from "expo-sharing";
import * as Print from "expo-print";
import { TYPE_META } from "./templates";
import { formatDate, formatOdo, formatLogLine } from "../format";

function stamp() {
  return new Date().toISOString().slice(0, 10);
}

function csvCell(value) {
  const s = value == null ? "" : String(value);
  if (/[",\n\r]/.test(s)) return `"${s.replace(/"/g, '""')}"`;
  return s;
}

function esc(value) {
  return String(value ?? "")
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}

function typeLabel(type) {
  return TYPE_META[type]?.label || type || "";
}

function slug(name) {
  return (
    String(name || "export")
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, "-")
      .replace(/^-|-$/g, "")
      .slice(0, 40) || "export"
  );
}

function allLogs(vehicles) {
  return vehicles
    .flatMap((v) =>
      (v.logs || []).map((l) => ({
        ...l,
        vehicle: v,
        vehicleName: v.name,
        vehicleType: typeLabel(v.type),
        vehicleModel: v.model || "",
      }))
    )
    .sort((a, b) => new Date(b.date) - new Date(a.date));
}

export function buildJson(vehicles) {
  return JSON.stringify(
    {
      app: "GreaseTrail",
      version: 1,
      exportedAt: new Date().toISOString(),
      vehicles,
    },
    null,
    2
  );
}

export function buildCsv(vehicles) {
  const header = ["Vehicle", "Type", "Model", "Date", "Odometer_km", "Category", "Service", "Values", "Note"];
  const rows = allLogs(vehicles).map((l) =>
    [
      l.vehicleName,
      l.vehicleType,
      l.vehicleModel,
      l.date,
      l.odometer,
      l.categoryName,
      l.service || "",
      formatLogLine(l.vehicle, l),
      l.note || "",
    ].map(csvCell)
  );
  return [header.join(","), ...rows].join("\n");
}

function pdfDetails(vehicle, log) {
  const parts = [];
  if (log.service) parts.push(`<div class="svc">${esc(log.service)}</div>`);
  const line = formatLogLine(vehicle, log);
  if (line) parts.push(`<div>${esc(line)}</div>`);
  if (log.note) parts.push(`<div class="note">${esc(log.note)}</div>`);
  return parts.join("") || "—";
}

export function buildPdfHtml(vehicles) {
  const single = vehicles.length === 1;
  const blocks = vehicles
    .map((v) => {
      const logs = [...(v.logs || [])].sort((a, b) => new Date(b.date) - new Date(a.date));
      const rows =
        logs.length === 0
          ? `<p class="empty">No log entries.</p>`
          : `<table>
              <thead>
                <tr>
                  <th>Date</th>
                  <th>km</th>
                  <th>Category</th>
                  <th>Details</th>
                </tr>
              </thead>
              <tbody>
                ${logs
                  .map(
                    (l) => `<tr>
                      <td>${esc(formatDate(l.date))}</td>
                      <td>${esc(formatOdo(l.odometer))}</td>
                      <td>${esc(l.categoryName)}</td>
                      <td>${pdfDetails(v, l)}</td>
                    </tr>`
                  )
                  .join("")}
              </tbody>
            </table>`;

      const heading = single
        ? ""
        : `<h2>${esc(v.name)}</h2>
        <p class="meta">${esc([v.model, typeLabel(v.type), `${formatOdo(v.odometer)} km`].filter(Boolean).join("  ·  "))}</p>`;

      return `<section>
        ${heading}
        ${rows}
      </section>`;
    })
    .join("");

  const title = single ? vehicles[0].name : "GreaseTrail";
  const sub = single
    ? [
        formatDate(new Date().toISOString()),
        vehicles[0].model,
        typeLabel(vehicles[0].type),
        `${formatOdo(vehicles[0].odometer)} km`,
      ]
        .filter(Boolean)
        .join("  ·  ")
    : `${formatDate(new Date().toISOString())}  ·  ${vehicles.length} vehicles`;

  return `<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8" />
  <style>
    body { font-family: -apple-system, Helvetica, Arial, sans-serif; color: #1A1A1A; background: #F5F5F0; padding: 28px; }
    h1 { font-size: 22px; letter-spacing: 0.08em; text-transform: uppercase; margin: 0 0 6px; }
    .sub { color: #8A8A84; font-size: 11px; letter-spacing: 0.12em; text-transform: uppercase; margin-bottom: 28px; }
    h2 { font-size: 16px; letter-spacing: 0.04em; text-transform: uppercase; margin: 28px 0 4px; }
    .meta { color: #8A8A84; font-size: 12px; margin: 0 0 12px; }
    .empty { color: #8A8A84; font-size: 13px; }
    .svc { font-weight: 600; margin-bottom: 2px; }
    .note { color: #8A8A84; margin-top: 2px; }
    table { width: 100%; border-collapse: collapse; }
    th { text-align: left; font-size: 10px; letter-spacing: 0.12em; text-transform: uppercase; color: #8A8A84; padding: 8px 8px 8px 0; border-bottom: 1px solid #E4E2DC; }
    td { font-size: 12px; padding: 8px 8px 8px 0; border-bottom: 1px solid #E4E2DC; vertical-align: top; }
    section { page-break-inside: avoid; }
  </style>
</head>
<body>
  <h1>${esc(title)}</h1>
  <p class="sub">${esc(sub)}</p>
  ${blocks || `<p class="empty">No vehicles.</p>`}
</body>
</html>`;
}

async function shareFile(uri, mimeType, dialogTitle) {
  const available = await Sharing.isAvailableAsync();
  if (!available) {
    Alert.alert("Share", "Sharing is not available on this device.");
    return;
  }
  await Sharing.shareAsync(uri, {
    mimeType,
    dialogTitle,
    UTI: mimeType === "application/pdf" ? "com.adobe.pdf" : mimeType === "text/csv" ? "public.comma-separated-values-text" : "public.json",
  });
}

async function writeCache(name, contents, encoding = FileSystem.EncodingType.UTF8) {
  const uri = `${FileSystem.cacheDirectory}${name}`;
  await FileSystem.writeAsStringAsync(uri, contents, { encoding });
  return uri;
}

export async function exportJson(vehicles) {
  const uri = await writeCache(`greasetrail-${stamp()}.json`, buildJson(vehicles));
  await shareFile(uri, "application/json", "Export JSON");
}

export async function exportCsv(vehicles) {
  const uri = await writeCache(`greasetrail-${stamp()}.csv`, buildCsv(vehicles));
  await shareFile(uri, "text/csv", "Export CSV");
}

export async function exportPdf(vehicles) {
  const label = vehicles.length === 1 ? slug(vehicles[0].name) : "all";
  const { uri } = await Print.printToFileAsync({ html: buildPdfHtml(vehicles) });
  const dest = `${FileSystem.cacheDirectory}greasetrail-${label}-${stamp()}.pdf`;
  if (uri !== dest) {
    try {
      await FileSystem.copyAsync({ from: uri, to: dest });
      await shareFile(dest, "application/pdf", "Export PDF");
      return;
    } catch (e) {
      // Fall through and share the print file as-is.
    }
  }
  await shareFile(uri, "application/pdf", "Export PDF");
}

export async function shareJsonFile(filename, contents, title) {
  const uri = await writeCache(filename, contents);
  await shareFile(uri, "application/json", title);
}
